### PR TITLE
Perform SSA variable renaming.

### PR DESCRIPTION
**Pull request by vext01**
Please check carefully. I've been working on this code for so long I feel I can barely see the wood for the trees!

There's a ykpack change associated with this PR. Coming soon.

We perform renaming in-place using a slightly modified version of the
algorithm from the book: Modern Compiler Implementation in Java, 2nd
edition, Andrew Appel.

We experimented with the algorithm from the SSA book by Jeremy Singer
but couldn't get it to work. I suspect there is an error in the
published algorithm, but I may be wrong.

For now this only renames variables for implemented TIR bytecodes, and
not yet in terminators (so return terminators currently always return
variable 0, which will need to be changed).

It may be possible to have the variables numbers start from zero in the
resulting bytecode, but I'm keen to get this in first.

Here are some example CFGs:

image

---

image

### CODE REVIEW

File: src/librustc_yk_sections/Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,8 @@ test = false
 [dependencies]
 rustc = { path = "../librustc" }
 rustc_yk_link = { path = "../librustc_yk_link" }
-ykpack = { git = "https://github.com/softdevteam/ykpack" }
+#ykpack = { git = "https://github.com/softdevteam/ykpack" }
+ykpack = { path = "../../../ykpack" }
```
Comment (vext01) [inline src/librustc_yk_sections/Cargo.toml:16]: This will need to be revised once the ykpack change is in.

**Comment by vext01** [comment]
I just realised that ykpack is now in the mono repo. We could do with that transition being complete before I raise the companion PR to this PR.

**Comment by ltratt** [comment]
> I just realised that ykpack is now in the mono repo. We could do with that transition being complete before I raise the companion PR to this PR.

The transition is complete isn't it?

File: src/librustc_data_structures/graph/dominators/mod.rs [hunk 2]
```diff
@@ -118,6 +118,15 @@ impl<Node: Idx> Dominators<Node> {
         self.immediate_dominators[node].unwrap()
     }
 
+    /// Find the children of a node in the dominator tree.
+    pub fn immediately_dominates(&self, node: Node) -> Vec<Node> {
```
Comment (ltratt) [inline src/librustc_data_structures/graph/dominators/mod.rs]: I don't know if this will be possible (it probably depends on the borrow checker), but it would almost certainly be faster to make this an iterator rather than a producer of a `Vec`. You could probably get away with `-> impl Iterator<Node>` or similar.

**Comment by vext01** [comment]
> The transition is complete isn't it?

I don't think so. We have an outstanding PR here:
https://github.com/softdevteam/yk/pull/1

File: src/librustc_yk_sections/emit_tir.rs [hunk 3]
```diff
@@ -29,16 +34,26 @@ use std::io::Write;
...
+use rustc_data_structures::graph::WithSuccessors;
 use ykpack;
 use ykpack::LocalIndex as TirLocal;
+use ykpack::BasicBlockIndex as TirBasicBlockIndex;
 use rustc_data_structures::fx::FxHashSet;
 
 const SECTION_NAME: &'static str = ".yk_tir";
 const TMP_EXT: &'static str = ".yk_tir.tmp";
 
+/// The pre-SSA return value variable. In the pre-SSA TIR, we need the return value to reside at an
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs]: The first half of this comment makes sense, but the second half is confusing. Possibly it's just because some sentences are in the wrong order?

File: src/librustc_yk_sections/emit_tir.rs [hunk 4]
```diff
@@ -57,25 +72,52 @@ struct ConvCx<'a, 'tcx, 'gcx> {
     /// Maps each block to the variable it defines. This is what Appel calls `A_{orig}`.
     block_defines: RefCell<IndexVec<BasicBlock, FxHashSet<TirLocal>>>,
     /// Monotonically increasing number used to give TIR variables a unique ID.
+    /// Note that a 0 is reserved for `PRE_SSA_RET_VAR`.
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs]: "a 0" -> "0"

File: src/librustc_yk_sections/emit_tir.rs [hunk 5]
```diff
@@ -57,25 +72,52 @@ struct ConvCx<'a, 'tcx, 'gcx> {
...
+        }
+    }
+
+    /// Make a SSA definition of all variables (before renaming) at the entry point. Call this
+    /// immediately after constructing a `RenameCx`.
+    ///
+    /// From the Appel book:
+    ///     "We consider the start node to contain an implicit definition of every variable,
+    ///     either because the variable may be a formal parameter or to represent the notion of
+    ///     a ← uninitialized without special cases.
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs]: I can't parse the part of the comment on this line.

File: src/librustc_yk_sections/emit_tir.rs [hunk 6]
```diff
@@ -262,6 +306,108 @@ fn insert_phi(block: &mut ykpack::BasicBlock, var: TirLocal, arity: usize) {
...
+    }
+
+    // FIXME rename variables in terminators.
+    fn rename(&mut self, doms: &Dominators<BasicBlock>, mir: &Mir,
+        blks: &mut Vec<ykpack::BasicBlock>, n: TirBasicBlockIndex)
+    {
+        let n_usize = n as usize;
+        // We have to remember the variables whose stacks we must pop from when we come back from
+        // recursion. These must be the variables *before* they were renamed.
+        let mut pop_later = Vec::new();
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs:370]: Can a variable appear more than once in here? Does it happen often that variables end up in here?

File: src/librustc_yk_sections/emit_tir.rs [hunk 7]
```diff
@@ -262,6 +306,108 @@ fn insert_phi(block: &mut ykpack::BasicBlock, var: TirLocal, arity: usize) {
...
+                }
+            }
+        }
+
+        let n_idx = BasicBlock::new(n_usize);
+        for y in mir.successors(n_idx) {
+            // "Suppose n is the jth predecessor of y".
+            let j = mir.predecessors_for(y).iter().position(|b| b == &n_idx).unwrap();
+            for st in &mut blks[y.as_usize()].stmts {
+                match st.phi_arg_mut(j) {
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs]: `if let Some(ref mut a) = st.phi_arg_mut(j) { ... } `

File: src/librustc_data_structures/graph/dominators/mod.rs (hunk 2 above)
Comment (vext01) [reply src/librustc_data_structures/graph/dominators/mod.rs]: There was a reason why I didn't do this, but I don't recall what it was. Let me experiment and come back to you.

File: src/librustc_yk_sections/emit_tir.rs (hunk 3 above)
Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs]: Can you expand on that -- it makes sense to me, but then I'm not a good critic for my own work.

Comment (ltratt) [reply src/librustc_yk_sections/emit_tir.rs]: I think you just need to try rephrasing it. I can't make sense of the comment as a whole at the moment.

File: src/librustc_yk_sections/emit_tir.rs (hunk 5 above)
Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs]: Something has gone wrong somewhere. That's supposed to be \bottom. Will fix.

File: src/librustc_yk_sections/emit_tir.rs (hunk 6 above)
Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs:370]: Every variable assignment in the block we are processing will push to this vector. The same variable may appear more than once if a block assigns to the same variable many times.

Comment (ltratt) [reply src/librustc_yk_sections/emit_tir.rs:370]: So this should probably be a set to avoid (I assume pointlessly) poping the same thing over and over again?

Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs:370]: If a variable is assigned to X times, then we need to pop X times.

The the algorithm recursively walks the dominator tree depth-first. When it sees an assignment to a variable, it makes a new SSA variable, rewrites the old variable. and pushes the new variable onto the stack of the corresponding old variable. When we are bubbling up from recursion, we have to revert to the previous SSA versions of the variables. So if a block assigned to a variable `z` twice, we'd need to revert to `z` as of two assignments ago, thus popping twice.

Hard to explain...

Comment (ltratt) [reply src/librustc_yk_sections/emit_tir.rs:370]: Got it.

File: src/librustc_yk_sections/emit_tir.rs (hunk 5 above)
Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs]: Actually I was wrong. This is verbatim from the book.

image

Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs]: maybe we can delimit the abstract algebra in some quotes for readability. I think the problem is `notion of a <- uninitialized` is hard to parse.

**Comment by vext01** [comment]
This should address the comments.

Untested. Building now...

File: src/librustc_yk_sections/emit_tir.rs (hunk 5 above)
Comment (ltratt) [reply src/librustc_yk_sections/emit_tir.rs]: Agreed, maybe this just needs a bit of rephrasing / reformatting.

Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs]: I added the quotes in the last commit. Does that work for you?

**Comment by vext01** [comment]
> Untested. Building now...

Our yk_tir test passes and manual inspection of the graphs show up nothing dodgy.

File: src/librustc_yk_sections/emit_tir.rs [hunk 8]
```diff
@@ -47,11 +47,11 @@ use rustc_data_structures::fx::FxHashSet;
...
 
-/// The pre-SSA return value variable. In the pre-SSA TIR, we need the return value to reside at an
-/// easily locatable variable index so that later we can easily convert from MIR's implicit return
-/// values, to TIR's explicit return values. In MIR, a return terminator implicitly returns
-/// variable 0. Implicit return values are not possible for TIR, as the actual value we need to
-/// return depends upon which SSA definition reaches the return terminator.
+/// The pre-SSA return value variable. In MIR, a return terminator implicitly returns variable
+/// zero. We can't do this in TIR because TIR is in SSA form and the variable we return depends
+/// upon which SSA variable reaches the terminator. So, during initial TIR lowering, we convert the
+/// implicit MIR terminator to an explicit TIR terminator returning variable index zero initially.
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs]: I think we can remove the last  "initially".

File: src/librustc_yk_sections/emit_tir.rs [hunk 9]
```diff
@@ -388,13 +388,12 @@ impl RenameCx {
...
             for st in &mut blks[y.as_usize()].stmts {
-                match st.phi_arg_mut(j) {
-                    Some(ref mut a) => {
-                        let i = self.stack[**a as usize].last().cloned().unwrap();
-                        **a = i;
-                    },
-                    None => (), // It wasn't a Phi. Do nothing.
+                if let Some(ref mut a) = st.phi_arg_mut(j) {
+                    // We only get here if `st` was a Phi.
+                    let i = self.stack[**a as usize].last().cloned().unwrap();
```
Comment (ltratt) [inline src/librustc_yk_sections/emit_tir.rs:397]: `let i =` is longer than `**a =`. Can we use the latter?

Comment (vext01) [reply src/librustc_yk_sections/emit_tir.rs:397]: We can, but I'd rather keep it as-is for parity with the algorithm in the book:

image

Comment (ltratt) [reply src/librustc_yk_sections/emit_tir.rs:397]: OK.

**Comment by vext01** [comment]
One commit to address your comment, another to give the instruction that does the SSA definitions a better name.

If you are happy, we can squash this ready for merge, but we cannot merge just yet. There's a yk change that I'm about to raise a PR for and that depends on the new tarball PR with cargo.

Sadly my stage 2 failed because I forgot to specify the config file on the command line and now I have to rebuild...

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
squashed. Now just waiting on the ykrustc archives. Once that's done, I'll flip the ykpack dependency over to yk.

**Comment by vext01** [comment]
This is now irrelevant.
